packages/core/src/components/toast/overlayToaster.tsx

```diff
@@ -146,21 +146,35 @@ export class OverlayToaster extends AbstractPureComponent<OverlayToasterProps, O
     };
 
     public show(props: ToastProps, key?: string) {
+        const options = this.createToastOptions(props, key);
+        const updatedExistingToast = this.maybeUpdateExistingToast(options, key);
+        if (updatedExistingToast) {
+            return options.key;
+        }
+
         if (this.props.maxToasts) {
             // check if active number of toasts are at the maxToasts limit
             this.dismissIfAtLimit();
         }
-        const options = this.createToastOptions(props, key);
+
+        this.updateToastsInState(toasts => [options, ...toasts]);
+        return options.key;
+    }
+
+    private maybeUpdateExistingToast(options: ToastOptions, key: string | undefined) {
+        if (key == null || this.isNewToastKey(key)) {
+            return false;
+        }
+
+        this.updateToastsInState(toasts => toasts.map(t => (t.key === key ? options : t)));
+        return true;
+    }
+
+    private updateToastsInState(getNewToasts: (toasts: ToastOptions[]) => ToastOptions[]) {
```

Comment:
I've noticed that the current abstraction with this function taking another function as an argument is making the code a bit harder for me to follow. It seems a bit premature to abstract it in this way.

Would it be possible to simplify by pulling the check for updating an existing toast to the top and using it directly in the conditional for dismissing? e.g.
```ts
public show(props: ToastProps, key?: string) {
    const options = this.createToastOptions(props, key);
    const isUpdatingExistingToast = key != null && !this.isNewToastKey(key);

    if (this.props.maxToasts && !isUpdatingExistingToast) {
        // check if active number of toasts are at the maxToasts limit
        this.dismissIfAtLimit();
    }

    this.setState(prevState => {
        const toasts = isUpdatingExistingToast
            ? // update a specific toast
              prevState.toasts.map(t => (t.key === key ? options : t))
            : // prepend a new toast
              [options, ...prevState.toasts];

        return { toasts, toastRefs: this.getToastRefs(toasts) };
    });
    return options.key;
}
```

Reply:
I agree that its a bit premature in the context of this PR, but with my FLUP change for queuing, it becomes more needed as we're doing these state updates in a couple of places where its not so easy to refactor. If you're down, lets leave this as is and maybe in the context of the other PR we can reevaluate?